### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by siewhui2023**
Implemented a Siamese Neural Network (SNN) to classify the ADNI dataset. It consists of pair generation to create training data, train-validation splitting (80/20), a CNN feature extractor, a Siamese network with Euclidean distance & contrastive loss to learn feature representations, and a Classifier model built on top of the pre-trained CNN. The Siamese network and Classifier achieve decent accuracy and loss but have some overfitting. The test accuracy is 70%+, showing good generalisation. 

Details of each file: 
dataset.py: Load all images and data preparation/preprocessing
module.py: Construction of the neural network, Siamese network & Classifier model, euclidean distance and contrastive loss
train.py: Train, save and plot the models
predict.py: Accuracy and loss (on test data) of the trained model
README.md: Consists of all the details (description of SNN, approach, results) 

Thank you!
Siew Hui Chun (47242467)

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Very good
- clear flow of information
- algorithm, training choices, usage explained well
- relevant plots/figures provided
- results followed with discussion/interpretation

Commit messages: OK, only a few

Code: 
- design is ok
- sufficiently commented
- algorithm seem to be impelmented correctly

Functionality/Performance: 
- both siamese and classifier seem to be performing well
- looking at the loss and accuracy plots, i think there's no overfitting. It's expected that the validation set will not show the same performance as the training set, and your implementation works good enough
- test set classification accuracy is only 70%

General comments:
- algorithm solves the problem apporpriately although test set performance is poor

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good, usage missing -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
